handle errors in zip_input_files

bertsky opened the pull request:
See [discussion on OCR-D/core#1240](https://github.com/OCR-D/core/pull/1240#discussion_r1721897464)

kba committed:
add config.OCRD_DOWNLOAD_INPUT

Commit:
Processor.zip_input_files: warning instead of exception for missing input files

Commit:
Processor.zip_input_files: introduce NonUniqueInputFile exception

Commit:
Processor.process_workspace: zip_input_files w/o require_first

Commit:
Merge remote-tracking branch 'origin/download-files-config-var' into new-processor-api-input-file-errors

Commit:
Processor.zip_input_files: introduce MissingInputFile exception and config.OCRD_MISSING_INPUT

bertsky commented:
@kba I tentatively merged #9 here.

@kba @MehmedGIT this effectively undoes some of your recent commits:
- https://github.com/OCR-D/core/commit/15a2c08a98ba4e0147c56dc89f23380252de0bb2
- https://github.com/OCR-D/core/commit/1476d95beee5a9a787995be81f6e9ce4db5779ac
- https://github.com/OCR-D/core/commit/64b356384be8155797ae6147f23447773992c2b0

Please review (cannot invite you directly).

MehmedGIT commented on src/ocrd/processor/base.py:
```diff
@@ -575,67 +605,66 @@ def zip_input_files(self, require_first=True, mimetype=None, on_error='skip'):
                         if on_error == 'skip':
                             ift[i] = None
                         elif on_error == 'first':
                             pass # keep first match
                         elif on_error == 'last':
                             ift[i] = file_
                         elif on_error == 'abort':
-                            raise ValueError(
-                                "Multiple '%s' matches for page '%s' in fileGrp '%s'." % (
-                                    mimetype, file_.pageId, ifg))
+                            raise NonUniqueInputFile(ifg, file_.pageId, mimetype)
                         else:
                             raise Exception("Unknown 'on_error' strategy '%s'" % on_error)
```
This should become a separate method. I see the same if checks are repeated below. The constant strings for the `on_error` strategy could also become an Enum class. The Enum class would also help with the validator and parser in the config.

bertsky replied:
> This should become a separate method. I see the same if checks are repeated below.

We used to have two discerning error messages here. (In one case, a specific MIME type was requested, like PAGE-only or image-only or JSON or whatever, but that was not unique. In another, no MIME type was requested, so the PAGE+images logic should apply but did not establish uniqueness via some PAGE file.)

How would you like to see this refactored into a function, exactly?

> The constant strings for the `on_error` strategy could also become an Enum class. 

Agreed.

> The Enum class would also help with the validator and parser in the config.

Remember that this concrete setting should not be decided by the user. The config has a different granularity level: what should happen to input files missing (or skipped because of being non-unique).

MehmedGIT replied:
> How would you like to see this refactored into a function, exactly?

Something like:

```python
def handle_errors(ift[i], file_, ifg: str, on_error: str):
  if on_error == 'skip':
      ift[i] = None
  elif on_error == 'first':
      pass # keep first match
  elif on_error == 'last':
      ift[i] = file_
  elif on_error == 'abort':
      raise NonUniqueInputFile(ifg, file_.pageId, None)
  else:
      raise Exception("Unknown 'on_error' strategy '%s'" % on_error)
```

or

```python
def handle_errors(file_, on_error: str, ifg: str) -> Union[None, file_]:
  if on_error == 'skip':
      return None
  elif on_error == 'first':
      return None
  elif on_error == 'last':
      return file_
  elif on_error == 'abort':
      raise NonUniqueInputFile(ifg, file_.pageId, None)
  else:
      raise Exception("Unknown 'on_error' strategy '%s'" % on_error)
```

The only difference I see is the warning logs content which could be preserved outside the method. If different things happened, say on skip and/or abort, then we no longer have the method option. Anyway, nothing important for the current phase of v3.

bertsky replied:
Ok, sure, we can pass in the differentiating part of the message for NonUniqueInputFile.

But this is problematic:

> ```python
> if on_error == 'skip':
>       return None
>   elif on_error == 'first':
>       return None
> ```

How would you call that function?

With skip, we should set the input file to None. With first, we should keep the previous OcrdFile instance.

MehmedGIT replied:
Right, the return version would not work and the other version requires passing some arguments for just checking the on_error state. Let's leave it as it is.

MehmedGIT commented on src/ocrd_utils/config.py:
```diff
@@ -141,6 +143,18 @@ def _ocrd_download_timeout_parser(val):
+config.add("OCRD_MISSING_INPUT",
+    description="How to deal with missing input files (for some fileGrp/pageId) during processing [SKIP|ABORT]",
+    default=(True, 'SKIP'),
+    validator=lambda val: val in ['SKIP', 'ABORT'],
```
What about FIRST and LAST? Or are these supposed to be only internals?

bertsky replied:
Does not apply here. Remember `first` and `last` are only for the internal problem of dealing with non-unique matches. We did not want to bother the user with this level.

Commit:
processor parameter decorator: no '{}' default (unnecessary)

Commit:
Processor: add error handling…

- introduce `config.OCRD_MISSING_OUTPUT` and
  catch exceptions during `process_page_file`:
  - `ABORT`: re-raise
  - `SKIP`: ignore and continue with next page
  - `COPY`: ignore and provide input PAGE-XML
    as output (with just `@PcGtsId` and `Metadata`
    added to simulate the processing step)

- introduce `config.OCRD_EXISTING_OUTPUT`:
  - `ABORT`: re-raise FileExistsError
  - `SKIP`: ignore and continue with next page
  - `OVERWRITE`: force overwriting the exact
    output files (instead of removing output files
    indiscriminately)

- :fire: remove `Workspace.overwrite_mode`,
  have `--overwrite` merely delegate to
  `config.OCRD_EXISTING_OUTPUT=OVERWRITE`

- introduce `--debug`, just delegate to
  `config.OCRD_MISSING_OUTPUT=ABORT`

- `cli.bashlib.input-files`: delegate everything
  to `ocrd_cli_wrap_processor` (for CLI handling)
  and `process_workspace` (for error handling),
  but override `process_page_file` to (never fail and)
  print bash-friendly strings for actual processing

- update tests, add `test_processor.test_run_output_missing`
  covering all `OCRD_MISSING_OUTPUT` options and the newly
  `OCRD_EXISTING_OUTPUT=SKIP`

bertsky commented:
f00ecda: this covers the remaining aspects of error handling on the processor level as I understand them:
- dealing with exceptions during the `process_page_file` call tree; acting on `config.OCRD_MISSING_OUTPUT` accordingly; `--debug` behaviour
- avoiding FileExistsError or dealing with it after `process_page_file` calls; acting on `config.OCRD_EXISTING_OUTPUT` accordingly; `--overwrite` behaviour

By @kba's request, I pushed this to the same PR so we can quickly go forward merging everything into `new-processor-api` for the next release. So please re-review!

MehmedGIT commented on src/ocrd_utils/config.py:
```diff
@@ -141,6 +143,30 @@ def _ocrd_download_timeout_parser(val):
+    default=(True, 'SKIP'),
+    validator=lambda val: val in ['SKIP', 'OVERWRITE', 'ABORT'],
+    parser=str)
+
```
It is not urgent, but I would also appreciate a more detailed description of each option. What happens when the output is missing and a COPY option is selected? What is being copied? It was maybe discussed in more detail somewhere but I have missed it. The existing output options are straightforward.

MehmedGIT replied:
The comment was supposed to go for the `OCRD_MISSING_OUTPUT` context.

MehmedGIT replied:
The discussion is here: https://github.com/bertsky/core/commit/f00ecda84717d274126a736b67e3ab29e5bae83d#r145600564

MehmedGIT commented:
Seems I [commented](https://github.com/bertsky/core/commit/f00ecda84717d274126a736b67e3ab29e5bae83d#r145600564) on a specific commit and that got lost and is not shown here.

kba commented on src/ocrd/processor/base.py:
```diff
@@ -55,9 +55,36 @@ class ResourceNotFoundError(FileNotFoundError):
+        self.pageId = pageId
+        self.mimetype = mimetype
+        self.message = (f"Could not find input file for fileGrp {fileGrp} "
+                        f"and pageId {pageId} under mimetype {mimetype or 'PAGE+image(s)'}")
```
Appreciate the specific exceptions. Maybe we should keep all of those in a separate `ocrd_utils.exceptions` before we release 3.0 stable.

bertsky replied:
Wouldn't that defeat package encapsulation (and possibly lead to circularity)? If I use stuff from `ocrd` which throws an exception, which I want to be able to handle, then that should be in `ocrd` as well.

MehmedGIT replied:
Is there a case where `ocrd` module code will be used inside `ocrd_utils.exceptions` and then that code is imported back to `ocrd` from `ocrd_utils`? IMO, no circularity should happen if implemented properly, i.e., just importing the required definitions or methods of a class instead of `*`. This said dot operations in Python are also more expensive in terms of performance.

```python
import workspace
workspace.add_file()
```

is less efficient than

```python
from workspace import add_file
add_file()
```

due to the extra attribute lookup overhead (usually bad inside loops).

kba replied:
Interesting about the import syntax. I prefer the latter anyway, good to hear that it's also faster.

> Wouldn't that defeat package encapsulation (and possibly lead to circularity)? If I use stuff from ocrd which throws an exception, which I want to be able to handle, then that should be in ocrd as well.

The dependency graph is `ocrd_utils -> ocrd_models -> ocrd_validators -> ocrd -> ocrd_network`. So unless `ocrd_utils.exceptions` imports something from one of the other packages (like the problem with in https://github.com/bertsky/core/pull/12), there won't be an issue, at least for these simple, builtin-derived exceptions. If we want e.g. `JsonSchemaDeprecationWarning` in there as well, then we'd have to move it up the package hierarchy, e.g. to `ocrd.exceptions`. Which would not work for `JsonSchemaDeprecationWarning` because that is used in `ocrd_validators`.

It is not important, we just have quite a few custom exceptions and I thought it would be sensible to have them in one place, as many python projects do. But let's put a pin into that discussion until we're ready to release, there's more pressing issues.

kba commented on src/ocrd/processor/base.py:
```diff
@@ -287,17 +314,19 @@ def process_workspace(self, workspace: Workspace) -> None:
                     # - ResourceNotFoundError → use ResourceManager to download (once), then retry
                     # - transient (I/O or OOM) error → maybe sleep, retry
                     # - persistent (data) error → skip / dummy / raise
                     input_files : List[Optional[Union[OcrdFile, ClientSideOcrdFile]]] = [None] * len(input_file_tuple)
```
and in the same vain, it might be useful to have `ocrd.typing` containing stuff like

```python
OcrdFileType = Union[OcrdFile, ClientSideOcrdFile]
OcrdPageType = Union[OcrdPage, PcGtsType]
VersionType =  Union[Tuple[int, int, int], Tuple[int, int, int, str]]
```

bertsky replied:
Yes, that would be useful. But why the extra module level `typing`?

MehmedGIT replied:
I think it was meant to be a separate file inside the ocrd module, not a separate module.

kba replied:
It doesn't have to be a separate file/module, just somewhere importable. But it has the same potential pitfalls with circular dependencies as in the other cases discussed, so let's focus on the functionality first, cleaner codebase and optimizations later.

kba commented on src/ocrd/processor/base.py:
```diff
@@ -621,21 +638,21 @@ def zip_input_files(self, require_first=True, mimetype=None, on_error='skip'):
                     ift[i] = file_
+        # Warn if no files found but pageId was specified, because that might be due to invalid page_id (range)
+        if self.page_id and not any(pages):
+            LOG.critical(f"Could not find any files for selected pageId {self.page_id}")
```
```suggestion
            LOG.critical(f"Could not find any files for selected pageId {self.page_id}.\ncompare '{self.page_id}' with the output of 'orcd workspace list-page'.")
```

kba commented on tests/processor/test_processor.py:
```diff
@@ -251,7 +251,7 @@ def ocrd_tool(self):
                     tuples = [(one.ID, two) for one, two in proc.zip_input_files(on_error='skip')]
                     assert ('foobar3', None) in tuples
-                    with self.assertRaisesRegex(Exception, "No PAGE-XML for page .* in fileGrp .* but multiple matches."):
+                    with self.assertRaisesRegex(Exception, "Could not determine unique input file"):
```
```suggestion
                    with self.assertRaisesRegex(NonUniqueInputFile, "Could not determine unique input file"):
```

kba commented on tests/processor/test_processor.py:
```diff
@@ -260,7 +260,7 @@ def ocrd_tool(self):
                     proc.input_file_grp = 'GRP1,GRP2'
                     proc.page_id = page_id
-                    with self.assertRaisesRegex(Exception, "Multiple PAGE-XML matches for page"):
+                    with self.assertRaisesRegex(Exception, "Could not determine unique input file"):
```
```suggestion
                    with self.assertRaisesRegex(NonUniqueInputFile, "Could not determine unique input file"):
```

kba reviewed (approved):
LGTM, only suggestions for cosmetics, functionality-wise good to go IMHO!

Commit:
ocrd_utils.config: add variables to module docstring

Commit:
improve docstrings, re-generate docs

bertsky committed:
Processor.zip_input_files: more verbose log msg

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

bertsky committed:
test_processor: test for specific exception

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

kba commented:
`tests/processor/test_processor.py` is missing an `from ocrd.processor import NonUniqueInputFile`.

bertsky commented:
> `tests/processor/test_processor.py` is missing an `from ocrd.processor import NonUniqueInputFile`.

oops – thanks!

Commit:
test_processor: fix missing import
